directives/meta.py

```diff
@@ -22,5 +35,63 @@ class AplusMeta(Directive):
         'introduction': directives.unchanged, # module introduction HTML
     }
 
+    # Valid date formats:
+    # 1. 'YYYY-MM-DD [hh[:mm[:ss]]]'
+    # 2. 'DD.MM.YYYY [hh[:mm[:ss]]]'
+    date_format = re.compile(r"^(\d\d\d\d-\d\d-\d\d|\d\d.\d\d.\d\d\d\d)( \d\d(:\d\d(:\d\d)?)?)?$")
```

Comment:
Finnish would use format `DD.MM.YYYY HH.MM.SS`, if you want to be correct. Though, I don't know if that is accepted somewhere else.

Reply:
We should at least accept the date formats that have been previously accepted. Otherwise, some old course might not work with the new format check. At a quick glance, the `toc_config.py` seems to responsible for parsing dates: https://github.com/apluslms/a-plus-rst-tools/blob/c8bbe542d22246ff882e2cea0d10ffe65d071293/toc_config.py#L130

Reply:
raphendyr: True. I think it is a separate discussion whether there is need for more date and time formats. Trak Y has separate RST macros for Finnish date texts such as `|tentti1| replace:: 18.12.2020 klo 09-12`.

Mankro: Yes, I duplicated the date format validation from `parse_date()` in `toc_config.py`. To prevent code duplication, it would be best to use only one date parser function. Currently function `parse_date()` cannot be imported in `directives/meta.py`, because `toc_config.py` has line 7: `import directives.meta`. I propose that `parse_date()` is refactored into a new file `lib/config_parsing.py` from which both `toc_config.py` and `directives/meta.py` can import it. It would be good to document the function properly. My proposal:

**lib/toc_config.py**
`
def parse_date(src, allow_empty=False):
    """Parses a date from natural text into and converts it to standard format.
    
    Parameters:
    src: string containing a date text in one of the formats:
         * 'date'
         * 'date time'
         where:
         * 'date' is either in format 'dd.mm.YYYY' or 'YYYY-mm-dd', where
           'dd' day of month (01..31), 'mm' is number of month (01..12) and
           'YYYY' is the year. For example: '25.09.2020' or '2020-09-25'.
         * 'time' is one of the following:
           'HH'       hour in 24-hour clock
           'HH:MM'    minutes
           'HH:MM:SS' seconds
           fallback: set to 12:00 noon

    allow_empty: if True, allows (src == None)

    Returns:
    Date string in one of the following formats:
    'YYYY-mm-dd HH'
    'YYYY-mm-dd HH:MM'
    'YYYY-mm-dd HH:MM:SS'
    """
    if allow_empty and not src:
        return None

    parts = src.split(u' ', 1)
    d = parts[0] # date
    t = parts[1] if len(parts) > 1 else '' # time

    if re.match(r'^\d\d.\d\d.\d\d\d\d$', d):
        # Date in format DD.MM.YYYY (Date, Month, Year)
        # -> convert to YYYY-MM-DD
        ds = d.split('.')
        d = ds[2] + u'-' + ds[1] + u'-' + ds[0]

    elif re.match(r'^\d\d\d\d-\d\d-\d\d$', d):
        # Date in format YYYY-MM-DD (Year, Month, Date):
        # accepted as it is
        pass
    else:
        raise SphinxError(u'Invalid date ' + d)

    if not re.match(r'^\d\d(:\d\d(:\d\d)?)?$', t):
        t = u'12:00'
    return d + u' ' + t
`

Reply:
Yes, `lib/config_parsing.py` is a good idea. And let's start by using the same date formats that have previously been supported in the directive options.

Reply:
I have tested that both `parse_date` in `toc_config.py` and `date_format` in `meta.py` support the same date formats. The following strings are accepted:
`    '01.01.2020',
    '01.01.2020 10',
    '01.01.2020 10:32',
    '01.01.2020 10:32:15',
    '2020-01-02',
    '2020-01-02 15',
    '2020-01-02 15:12',
    '2020-01-02 15:12:11'`

An empty string is not accepted in those aplusmeta directive options that require a date. Reference: `toc_config.py` lines 261-269.
https://github.com/apluslms/a-plus-rst-tools/blob/c8bbe542d22246ff882e2cea0d10ffe65d071293/toc_config.py#L261
I will add a comment on this in the regexp variable `date_format` in `meta.py` and also improve the readability of the regexp by splitting it on several lines with comments.